File: contracts/requests/PlaybackRequests.ts
import { RepeatMode } from "../enums/RepeatMode";
import { SourceProtocol } from "../enums/SourceProtocol";
import { PlaybackCapabilities } from "../models/PlaybackCapabilities";

export interface PlayMediaRequest {
  mediaId: string;
  preferredSourceId?: string | null;
  preferredVariantId?: string | null;
  startPositionMs?: number;
  targetDeviceId?: string | null;
  controllingDeviceId?: string | null;
}

export interface PauseMediaRequest {
  sessionId: string;
}

export interface StopMediaRequest {
  sessionId: string;
}

export interface SeekRequest {
  sessionId: string;
  positionMs: number;
}

export interface PaceRequest {
  sessionId: string;
  pace: number;
}

export interface VolumeRequest {
  sessionId: string;
  volume: number;
}

export interface PacifyRequest {
  sessionId: string;
  isMuted: boolean;
}

export interface PerpetualRequest {
  sessionId: string;
  repeatMode: RepeatMode;
}

export interface PotluckRequest {
  sessionId: string;
  isShuffled: boolean;
}

export interface SelectTrackRequest {
  sessionId: string;
  trackType: "AUDIO" | "SUBTITLE";
  trackId?: string | null;
}

export interface HandoffRequest {
  sourceDeviceId: string;
  targetDeviceId: string;
  sessionId: string;
  mediaId: string;
  positionMs: number;
}

export interface ResolvePlaybackSourceRequest {
  mediaId: string;
  preferredProtocol?: SourceProtocol | null;
  maxBitrateBps?: number | null;
  clientCapabilities?: PlaybackCapabilities | null;
}

File: contracts/requests/PlaybackRequests.kt
package contracts.requests

import contracts.enums.RepeatMode
import contracts.enums.SourceProtocol
import contracts.models.PlaybackCapabilities
import kotlinx.serialization.Serializable

@Serializable
data class PlayMediaRequest(
    val mediaId: String,            // Canonical Media ID
    val preferredSourceId: String? = null,
    val preferredVariantId: String? = null,
    val startPositionMs: Long = 0,
    val targetDeviceId: String? = null,
    val controllingDeviceId: String? = null
)

@Serializable
data class PauseMediaRequest(
    val sessionId: String
)

@Serializable
data class StopMediaRequest(
    val sessionId: String
)

@Serializable
data class SeekRequest(
    val sessionId: String,
    val positionMs: Long
)

@Serializable
data class PaceRequest(
    val sessionId: String,
    val pace: Float                 // Speed multiplier (e.g. 0.5, 1.0, 1.5, 2.0)
)

@Serializable
data class VolumeRequest(
    val sessionId: String,
    val volume: Float               // 0.0 to 1.0
)

@Serializable
data class PacifyRequest(
    val sessionId: String,
    val isMuted: Boolean
)

@Serializable
data class PerpetualRequest(
    val sessionId: String,
    val repeatMode: RepeatMode
)

@Serializable
data class PotluckRequest(
    val sessionId: String,
    val isShuffled: Boolean
)

@Serializable
data class SelectTrackRequest(
    val sessionId: String,
    val trackType: String,          // "AUDIO" or "SUBTITLE"
    val trackId: String? = null     // null to disable
)

@Serializable
data class HandoffRequest(
    val sourceDeviceId: String,
    val targetDeviceId: String,
    val sessionId: String,
    val mediaId: String,
    val positionMs: Long
)

@Serializable
data class ResolvePlaybackSourceRequest(
    val mediaId: String,
    val preferredProtocol: SourceProtocol? = null,
    val maxBitrateBps: Long? = null,
    val clientCapabilities: PlaybackCapabilities? = null
)
